vscode/src/repository/repo-metadata-from-git-api.ts

```diff
@@ -0,0 +1,75 @@
+export interface RepoMetaData {
+    owner: string
+    repoName: string
+    repoVisibility: 'public' | 'private'
+}
+
+export class RepoMetadatafromGitApi {
+    // This class is used to get the metadata from the gitApi.
+    // It is primarily meant to get the
+    private static instance: RepoMetadatafromGitApi | null = null
+    private cache = new Map<string, RepoMetaData | undefined>()
+
+    private constructor() {}
+
+    public static getInstance(): RepoMetadatafromGitApi {
+        if (!RepoMetadatafromGitApi.instance) {
+            RepoMetadatafromGitApi.instance = new RepoMetadatafromGitApi()
+        }
+        return RepoMetadatafromGitApi.instance
+    }
+
+    public async getRepoMetadataUsingGitUrl(gitUrl: string): Promise<RepoMetaData | undefined> {
+        if (this.cache.has(gitUrl)) {
+            return this.cache.get(gitUrl)
+        }
+        const repoMetaData = await this.metadataFromGit(gitUrl)
+        this.cache.set(gitUrl, repoMetaData)
+        return repoMetaData
+    }
+
+    private async metadataFromGit(gitUrl: string): Promise<RepoMetaData | undefined> {
+        const ownerAndRepoName = this.parserOwnerAndRepoName(gitUrl)
+        if (!ownerAndRepoName) {
+            return undefined
+        }
+        const repoMetaData = await this.queryGitHubApi(ownerAndRepoName.owner, ownerAndRepoName.repoName)
+        return repoMetaData
+    }
+
+    private async queryGitHubApi(owner: string, repoName: string): Promise<RepoMetaData | undefined> {
+        const apiUrl = `https://api.github.com/repos/${owner}/${repoName}`
+        try {
+            const response = await fetch(apiUrl)
+            // For private repos without token the api returns 404
+            // Return the private repo visibility
+            if (response.status === 404) {
+                return {
+                    owner,
+                    repoName,
+                    repoVisibility: 'private',
+                }
+            }
+            if (!response.ok) {
+                return undefined
+            }
+            const repoData = await response.json()
+            return {
+                owner,
+                repoName,
+                repoVisibility: repoData.private ? 'private' : 'public',
+            }
+        } catch (error) {
+            return undefined
+        }
+    }
```

Comment:
```suggestion
    private async queryGitHubApi(owner: string, repoName: string): Promise<RepoMetaData | undefined> {
        const apiUrl = `https://api.github.com/repos/${owner}/${repoName}`
        const metadata = { owner, repoName, repoVisibility: false}
        try {
            const response = await fetch(repositoryUrl, { method: 'HEAD' })
            metadata.repoVisibility = response.ok
            return metadata
        } catch (error) {
            return undefined
        }
    }
```
I think we can just check if the response returns `ok` here since it returns false if 404?

<img width="753" alt="image" src="https://github.com/sourcegraph/cody/assets/68532117/ca573d83-1254-4d25-adaf-dfe4feacf3cd">

Example from https://github.com/sourcegraph/cody/blob/44ae4407980ee21ae0793143bd92c0d137810bd7/vscode/src/repository/git-extension-api.ts#L110-L127

Reply:
Thanks for the suggestion, looks much cleaner now.
In addition to response.ok, I am also checking `repoData.private` as per [github api documentation](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#create-an-organization-repository).
`metadata.isPublic = response.ok && repoData.private === false `